### DISCUSSION TITLE
How to add extensions with docker?

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#16989] [Archive]
I’ve set up directus with docker on my local machine. On previous projects I’ve used the CLI to set up directus, but according to [this comment](https://github.com/directus/docs/issues/247#issuecomment-1366484123), I’ll use docker compose in the future instead.

Let’s say I want to install [this extension](https://github.com/dimitrov-adrian/directus-extension-editorjs-interface). Previously I have simply ran `npm install directus-extension-editorjs`. But since using `docker compose`, I don’t know how to install or use it?

Do I have to write my own `Dockerfile` where the extension is installed/downloaded and copied to the right folder?

Is there a more convenient way to do this (like it was before)?

**Comment by br41nslug** [comment]
I don't have an easy solution for this one im afraid, loading extra npm packages into a docker is tricky. I personally prefer to stick to the original local extension for the time being. There are currently 3 approaches im aware of:
- Create a custom dockerfile with the extensions included (something like this should do but it's still a hassle)
  ```Dockerfile
  FROM directus/directus:9.22.1
  RUN npm install directus-extension-editorjs other-extensions
  ```
- Build the extension locally and mount it as volume into the right folder. https://github.com/dimitrov-adrian/directus-extension-editorjs-interface#building-locally-and-contributing
- Or sticking with the npm installation route, documentation was removed but will still work, until there are more user friendly solutions for this (we're working on it with the upcoming Marketplace)

**Reply by northern-mindset** [reply]
For truth seekers from Directus v10 era, here is a working example of a custom `Dockerfile`.

```Dockerfile
FROM directus/directus:10.2.1
USER root
RUN corepack enable \
  && corepack prepare pnpm@8.1.1 --activate \
  # Currently required, we'll probably address this in the base image in future release
  && chown node:node /directus

USER node
RUN pnpm install directus-extension-group-modal-interface directus-extension-api-trigger-interface directus-extension-board-layout directus-extension-computed-interface directus-extension-grid-layout directus-extension-inline-form-interface
```
from https://discord.com/channels/725371605378924594/1113799436087336971/1113863057484038206

> The project/docker image now use pnpm instead of npm to manage packages for starter, hence why npm doesn't work

**Reply by paescuj** [reply]
More up-to-date example here (for adding packages used in Flows, but it's the same for extension packages): https://docs.directus.io/self-hosted/docker-guide.html#adding-packages-to-use-in-flows-scripts

**Comment by br41nslug** [comment]
While im partial to the second approach i think think the first is the way to go for most people. To make that step as easiest as possible i'd use the following steps (all in the same directory):
- create the `Dockerfile` as described above
- edit your `docker-compose` file to use the Dockerfile instead of an existing image
  do this by replacing `image: directsus/directus:9.22.1` with `build: .`
- now when running `docker compose up -d` it should build the image to use
- when making changes to extensions or updating directus you may have to run `docker compose rebuild` for the image to be recreated

**Reply by kyuumeitai** [reply]
You can take this approach without a `Dockerfile`, you just have to  change the `command` section of your `docker-compose` service. I'm a total noob but my implementation looks like this:

      directus:
          image: directus/directus:9.22.4
          volumes:
            - ./apps/directus/extensions:/directus/extensions
            - ./apps/directus/snapshots:/directus/snapshots
            - ./apps/directus/uploads:/directus/uploads
          command: >
            sh -c "
              npm install directus-extension-editorjs
              npx directus schema snapshot ./snapshots/\"$(date \"+%F\")\"-snapshot-\"$(date \"+%s\")\".yaml
              npx directus database migrate:latest
              npx directus bootstrap && npx directus start
            "
      
and then the rest of your service definition (network, etc.)

Hope helps someone and if you guys have suggestions I will be happy to receive them.

**Reply by hodanli** [reply]
is this tested. when i tried no additional file created in directus/extensions. and no function added in directus ui.

**Reply by northern-mindset** [reply]
> i think the first is the way to go for most people. To make that step as easiest as possible i'd use the following steps

Will it ever get into the documentation? I found the current documentation to be quite obscure in this regard.

**Comment by ghost** [comment]
@br41nslug According to your [answer above](https://github.com/directus/directus/discussions/16989#discussioncomment-4570725), I’ve found a way to make your second approach more convenient.

By adding the following parts to `package.json` and then running `pnpm install`, the extension will be copied automatically to the right place. If we later want to update the package it will also be automatically copied to the right folder.

```json
{
    "scripts": {
        "postinstall": "mkdir -p extensions/interfaces/directus-extension-editorjs && cp node_modules/directus-extension-editorjs/dist/index.js extensions/interfaces/directus-extension-editorjs",
        "postupdate": "pnpm run postinstall"
    },
    "dependencies": {
        "directus-extension-editorjs": "^1.5.0"
    }
}

```

Assuming that a volume for `extensions` exists in the `docker-compose.yaml` file:

```yaml
        volumes:
            - ./uploads:/directus/uploads
            - ./extensions:/directus/extensions
```

**Reply by br41nslug** [reply]
I'd recommend one little change there (from my own workflow). Mounting the extension folder directly instead of overwriting the whole `/extensions` mount which has caused some edgecases in the past.
```yaml
        volumes:
            - ./uploads:/directus/uploads
            - ./node_modules/directus-extension-editorjs/dist:/directus/extensions/interfaces/directus-extension-editorjs
```
Leaving the script for moving files around unneeded 😄

**Reply by ghost** [reply]
Yes, that’s much easier. Thanks ;)

**Comment by franzpro** [comment]
Hi,

i tried to install an extension via: 
"npm install directus-extension-editorjs"

-> 
npm ERR! code EUNSUPPORTEDPROTOCOL
npm ERR! Unsupported URL Type "workspace:": workspace:*

no matter which extension. It shows the same Error. 

I also tried to install it via the docker file:
command: >
      sh -c "
        npm install directus-extension-editorjs
        "

I get for both variants the same error message. does anyone have an idea what this could be?

Directus V9.25.1
npm V9.5.1

thx
Franz

**Reply by rijkvanzanten** [reply]
@franzpro Can you try like this? https://github.com/directus/directus/issues/18081#issuecomment-1500089734

**Reply by franzpro** [reply]
Hi,

thank you. 

after: 

```
USER root
corepack enable 
corepack prepare pnpm@8.1.1 --activate
chown node:node /directus
```

Now it shows after:

```
USER node
pnpm install directus-extension-editorjs
->
 WARN  deprecated keyv-memcache@1.3.3: This project has been renamed to @keyv/memcache which will be updated moving forward.
 WARN  deprecated json2csv@5.0.7: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
 WARN  deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
 WARN  deprecated @types/cssnano@5.1.0: This is a stub types definition. cssnano provides its own type definitions, so you do not need this installed.
 WARN  deprecated har-validator@5.1.5: this library is no longer supported
 WARN  deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/
math-random for details.
 WARN  deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
 WARN  deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
 WARN  deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-U
S/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
Already up to date
Progress: resolved 992, reused 959, downloaded 0, added 0, done

devDependencies: skipped because NODE_ENV is set to production

 WARN  Issues with peer dependencies found
.
└─┬ @directus/api 9.25.1
  └─┬ @directus/extensions-sdk 9.25.1
    └─┬ rollup-plugin-styles 4.0.0
      └── ✕ unmet peer rollup@^2.63.0: found 3.19.1 in @directus/extensions-sdk
Done in 5.2s
```

---

I have no idea what the problem is. 

thx
Franz

**Reply by rijkvanzanten** [reply]
> I have no idea what the problem is.

Nothing! Seeing some warnings, but no errors. `Done in 5.2s` means all systems are good 👍🏻

**Reply by franzpro** [reply]
ok, probaby then i run in another problem, because the extension is not working... i installed two extensions like above: (with warnungs..)
"directus-extension-field-actions" and "directus-extension-editorjs"

Is there an way to proof which extension is loaded in directus?

thx
Franz

**Reply by aiac** [reply]
> Hi,
> 
> i tried to install an extension via: "npm install directus-extension-editorjs"
> 

Had the same issue with Direcuts 9.26. Replaced `npm install` with `yarn add`, and everything went well.

**Reply by franzpro** [reply]
Hi, 

thx. The installation also runs here with "yarn add" but the extension is unfortunately still not active. 
Which user did you use to execute the command? Are you also using Docker? 

I am still looking for what this could be.

**Comment by anutator** [comment]
**pnpm** doesn't install extensions to "**extensions**" folder. It installs extensions to **node_modules/.pnpm** folder. Besides if I don't delete package.json "pnpm install..." tries to reinstall many node_modules and makes docker image twice bigger.

**Reply by northern-mindset** [reply]
I also noticed that the image becomes twice bigger…

**Comment by anutator** [comment]
I don't know the best method, but I get it work in helm chart. I make symlinks to `node_modules/<name_of_directus_extension>` when pod is started and subdirs in `extensions` directory are created. You only have to change the list of extensions in ENV (not ARG).

Moving package.json in Dockerfile prevents reinstalling many node_modules.

```dockerfile
FROM directus/directus:10.3.0
....
USER root
RUN apk add --update --no-cache --allow-untrusted tzdata jq
RUN corepack enable

USER node

RUN corepack prepare pnpm@latest --activate

ENV EXTENSIONS="directus-extension-display-link directus-extension-computed-interface"

RUN mv package.json 111 \
 && pnpm config set registry $COREPACK_NPM_REGISTRY \
 && pnpm install $EXTENSIONS \
 && mv 111 package.json
```

Add additional postStart command to `deployment.yml` in helm chart:

```yaml
          lifecycle:
            postStart:
              exec:
                command:
                - "/bin/sh"
                - "-c"
                - >
                  until [ -d /directus/extensions/displays ];
                  do sleep 1;
                  done;
                  sleep 1;
                  for ext in $EXTENSIONS;
                  do ln -s /directus/node_modules/$ext/dist extensions/$(jq -r '."directus:extension".type' /directus/node_modules/$ext/package.json)s/$ext;
                  done
```
